### PR TITLE
Evolution operator

### PR DESCRIPTION
**Pull request by lillian542**
**Context:**

- `Exp` class does not have `generator` method or `parameter_frequency` property defined, even though these could be defined for some instances of `Exp`.
- For instances of `Exp` where the base operator includes parameters (e.g. `0.5*qml.PauliX(0)`, or any instance of `qml.Hamiltonian`), the operator has `num_params > 1`, and standard functions for differentiation via parameter shift, such as `qml.generator(op)`, don't work.

**Description of the Change:**

- Add `generator` and `parameter_frequency` to `Exp` class in cases where it is defined (when the coefficient has no real component and the base operator is Hermitian)
- Add `Evolution` operator defined to be of the form exp(ixG) with a single, trainable parameter x. This allows differentiation _with regards to the parameter x_, even in instances where the base operator G contains additional parameters.

### CODE REVIEW

File: pennylane/ops/op_math/exp.py [hunk 1]
```diff
@@ -271,3 +279,149 @@ def simplify(self):
...
+
+    Returns:
+       :class:`Exp`: A :class`~.operation.Operator` representing an operator exponential, of the form exp(-ixG).
+
+    **Use Details**
+
+    In contrast to the general Exp class, the Evolution operator is constrained to a single trainable
+    parameter, allowing it to be differentiated with regard to the evolution parameter even if the base
+    operator contains additional parameters.
+
```
Comment (lillian542) [inline pennylane/ops/op_math/exp.py:380]: I don't know if this is clear or helpful. Does it make sense? Is it confusing and out of context?

File: pennylane/ops/op_math/exp.py [hunk 2]
```diff
@@ -271,3 +279,149 @@ def simplify(self):
...
+    def __init__(self, generator, param, do_queue=True, id=None):
+        super().__init__(generator, coeff=1j * param, do_queue=do_queue, id=id)
+        self._name = "Evolution"
+        self.param = param
+
+    @property
+    def parameters(self):
+        """Parameter that the operator depends on."""
+        return [self.param]  # should this be the parameter or the coefficient?
+
```
Comment (lillian542) [inline pennylane/ops/op_math/exp.py:424]: Would like feedback regarding ^this^, as I'm not sure I have a good overview of what op.parameters is used for

File: pennylane/ops/op_math/exp.py [hunk 3]
```diff
@@ -280,3 +288,153 @@ def simplify(self):
...
+    If the base operator is Hermitian, then the gate can be used in a circuit,
+    though it may not be supported by the device and may not be differentiable.
+
+    >>> @qml.qnode(qml.device('default.qubit', wires=1))
+    ... def circuit(x):
+    ...     Evolution(qml.PauliX(0), -0.5 * x)
+    ...     return qml.expval(qml.PauliZ(0))
+    >>> print(qml.draw(circuit)(1.23))
+    0: ──Exp(0.00-0.61j X)─┤  <Z>
```
Comment (albi3ro) [inline pennylane/ops/op_math/exp.py:420]: Might be worth improving the `label` method so that it doesn't show the nonexistent real component.

File: pennylane/ops/op_math/exp.py [hunk 4]
```diff
@@ -280,3 +288,153 @@ def simplify(self):
...
+        + (1.0) [Z0 X1]
+
+        """
+        if np.real(self.coeff) != 0 or not self.base.is_hermitian:
+            raise GeneratorUndefinedError(
+                f"Exponential with coefficient {self.coeff} and base operator {self.base} does not have a generator."
+            )
+        return self.base
+
+    @property
```
Comment (albi3ro) [inline pennylane/ops/op_math/exp.py:314]: If this code is just duplicated from `Operation`, we could just make `Exp` inherit from `Operation` instead of `Operator` and simply inherit the code.

File: pennylane/ops/op_math/exp.py (hunk 1 above)
Comment (albi3ro) [reply pennylane/ops/op_math/exp.py:389]: Makes sense.  Phrasing could probably be polished a little.  Maybe breaking it up into shorter sentences.

File: pennylane/ops/op_math/exp.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/ops/op_math/exp.py:437]: Looks perfect.

File: pennylane/ops/op_math/exp.py (hunk 3 above)
Comment (lillian542) [reply pennylane/ops/op_math/exp.py:420]: done :)

File: pennylane/ops/op_math/exp.py (hunk 4 above)
Comment (lillian542) [reply pennylane/ops/op_math/exp.py:314]: It's identical to the code from Operation, but I wasn't sure how to make it inherit that without breaking things. Right now `Exp` inherits from `SymoblicOp` inherits from `Operator`. What would you suggest?

File: pennylane/ops/op_math/__init__.py [hunk 5]
```diff
@@ -54,7 +54,7 @@
 from .adjoint_class import Adjoint
 from .adjoint_constructor import adjoint
 from .controlled_class import Controlled, ControlledOp
-from .exp import exp, Exp
+from .exp import exp, Exp, Evolution
```
Comment (Jaybsoni) [inline pennylane/ops/op_math/__init__.py:57]: Add Evolution to the list of symbolic classes above so that it shows up in documentation 👍🏼

File: pennylane/ops/op_math/exp.py [hunk 6]
```diff
@@ -181,6 +184,11 @@ def has_decomposition(self):
                 )
         return math.real(self.coeff) == 0 and is_pauli_word(self.base)
 
+    @property
+    def inverse(self):
+        """Setting inverse not defined for Exp, so inverse is always False"""
+        return False
```
Comment (Jaybsoni) [inline pennylane/ops/op_math/exp.py:190]: Why was this needed?

File: pennylane/ops/op_math/exp.py [hunk 7]
```diff
@@ -280,3 +288,162 @@ def simplify(self):
...
+
+        >>> U.generator()
+          (0.5) [Y0]
+        + (1.0) [Z0 X1]
+
+        """
+        if np.real(self.coeff) != 0 or not self.base.is_hermitian:
+            raise GeneratorUndefinedError(
+                f"Exponential with coefficient {self.coeff} and base operator {self.base} does not have a generator."
+            )
```
Comment (Jaybsoni) [inline pennylane/ops/op_math/exp.py:311]: I wonder if this condition is sufficient? Suppose `base = qml.s_prod(1j, qml.Identity(wires=0))` and `coeff=1`. In this case both conditions are True, but clearly the generator should be Identity for this exponential operator.

This might be resolved by using `qml.simplify`, but could add unwanted overhead.

File: pennylane/ops/op_math/exp.py [hunk 8]
```diff
@@ -280,3 +288,162 @@ def simplify(self):
...
+        return self.base
+
+    @property
+    def parameter_frequencies(self):
+        r"""Returns the frequencies for each operator parameter with respect
+        to an expectation value of the form
+        :math:`\langle \psi | U(\mathbf{p})^\dagger \hat{O} U(\mathbf{p})|\psi\rangle`.
+
+        These frequencies encode the behaviour of the operator :math:`U(\mathbf{p})`
```
Comment (Jaybsoni) [inline pennylane/ops/op_math/exp.py:320]: missing `\hat{}` on U

File: pennylane/ops/op_math/exp.py [hunk 9]
```diff
@@ -280,3 +288,162 @@ def simplify(self):
...
+            except GeneratorUndefinedError as e:
+                raise ParameterFrequenciesUndefinedError(
+                    f"Operation {self.name} does not have parameter frequencies defined."
+                ) from e
+
+            with catch_warnings():
+                filterwarnings(
+                    action="ignore", message=r".+ eigenvalues will be computed numerically\."
+                )
+                eigvals = qml.eigvals(gen)
```
Comment (Jaybsoni) [inline pennylane/ops/op_math/exp.py:361]: are we manually raising the warning that is being caught in here?

File: pennylane/ops/op_math/exp.py [hunk 10]
```diff
@@ -280,3 +288,162 @@ def simplify(self):
...
+
+            eigvals = tuple(np.round(eigvals, 8))
```
Comment (Jaybsoni) [inline pennylane/ops/op_math/exp.py:363]: Oh boy, I'm not sure if rounding so implicitly is a good idea. Could we let the users select the precision? They should be aware of any rounding / changes in precision happening in the pipeline to ensure error calculations / assumptions are not distorted by manually rounding.

File: pennylane/ops/op_math/exp.py [hunk 11]
```diff
@@ -280,3 +288,162 @@ def simplify(self):
...
+        )
+
+
+class Evolution(Exp):
+
+    r"""Create an exponential operator that defines a generator, of the form :math:`e^{ix\hat{G}}`
+
+    Args:
+        base (~.operation.Operator): The Operator to be used as a Generator, G.
+        param=1 (Number): The evolution parameter, x. This parameter is not expected to have
```
Comment (Jaybsoni) [inline pennylane/ops/op_math/exp.py:378]:
```suggestion
        param (float): The evolution parameter, x. This parameter is not expected to have
```

File: tests/ops/op_math/test_exp.py [hunk 12]
```diff
@@ -577,3 +581,123 @@ def test_draw_integration(self):
         qml.drawer.tape_text(tape)
 
         assert "0: ──Exp─┤  "
+
+
+class TestDifferentiation:
```
Comment (Jaybsoni) [inline tests/ops/op_math/test_exp.py:586]: Can we add tests which check that the gradient is compute correctly (in general and for parameter shift). I am not 100% sure, but if we can use parameter shift and also provide the input parameter in a different interface (eg. torch or jax). Might be worth adding a test for that as well.

**Review by Jaybsoni — COMMENTED** [review]
Nice work so far! Made a few comments, let me know once they are addressed and I can give it another look 👍🏼

File: pennylane/ops/op_math/exp.py (hunk 6 above)
Comment (lillian542) [reply pennylane/ops/op_math/exp.py:190]: `qml.generator(op)` checks it, so if we want it to behave like other classes in terms of accessing the generator, it needs to have this property until inverse is removed everywhere

File: pennylane/ops/op_math/exp.py (hunk 7 above)
Comment (lillian542) [reply pennylane/ops/op_math/exp.py:311]: That is a good point, that would raise a GeneratorUndefinedError. If overhead is a concern, what if you don't simplify by default, but only to double-check? I.e.

```
  if np.real(self.coeff) == 0 and self.base.is_hermitian:
      return self.base

  simplified_op = qml.simplify(self)
  if np.real(simplified_op.coeff) == 0 and simplified_op.base.is_hermitian:
      return simplified_op.base
  
  raise GeneratorUndefinedError(
          f"Exponential with coefficient {self.coeff} and base ...
```

That way you only end up calling `qml.simplify` in instances where it's relevant, or where it's going to raise an error anyway.

File: pennylane/ops/op_math/exp.py (hunk 4 above)
Comment (lillian542) [reply pennylane/ops/op_math/exp.py:314]: Would it be okay to move it all the way up to Operator so both this and Operation can inherit it?

File: pennylane/ops/op_math/exp.py (hunk 10 above)
Comment (albi3ro) [reply pennylane/ops/op_math/exp.py:363]: This is because `eigvals_to_frequencies` has a caching.  Therefore, we make them immutable and with fewer significant figures.

File: pennylane/ops/op_math/exp.py (hunk 6 above)
Comment (albi3ro) [reply pennylane/ops/op_math/exp.py:190]: We could update `generator` to use `getattr(op, "inverse", False)` instead of `op.inverse`.

Comment (lillian542) [reply pennylane/ops/op_math/exp.py:190]: Is it correct that if we change it so `Exp` inherits from `Operation`, we need to overwrite this (and a couple other things) anyway? Since we don't want to have inversion on `Exp`.

File: pennylane/ops/op_math/exp.py [hunk 13]
```diff
@@ -280,3 +288,162 @@ def simplify(self):
...
+        For more details on this relationship, see :func:`.eigvals_to_frequencies`.
+        """
+        if self.num_params == 1:
+            # if the operator has a single parameter, we can query the
+            # generator, and if defined, use its eigenvalues.
+            try:
+                gen = qml.generator(self, format="observable")
+            except GeneratorUndefinedError as e:
+                raise ParameterFrequenciesUndefinedError(
+                    f"Operation {self.name} does not have parameter frequencies defined."
```
Comment (obliviateandsurrender) [inline pennylane/ops/op_math/exp.py:354]:
```suggestion
                    f"Parameter frequencies could not be computed for the operation {self.name} as its generator is not defined."
```
Same for the error below.

File: pennylane/ops/op_math/exp.py [hunk 14]
```diff
@@ -280,3 +288,162 @@ def simplify(self):
...
+        >>> gen = qml.generator(U, format="observable")
+        >>> gen
+        PauliX(wires=[0])
+        >>> gen_eigvals = qml.eigvals(gen)
+        >>> qml.gradients.eigvals_to_frequencies(tuple(gen_eigvals))
+        (2,)
+
+        For more details on this relationship, see :func:`.eigvals_to_frequencies`.
+        """
+        if self.num_params == 1:
```
Comment (obliviateandsurrender) [inline pennylane/ops/op_math/exp.py:347]: For non-parameterized gates, does having parameterized frequencies makes any sense? If not, I think we should raise a separate error for them.

File: pennylane/ops/op_math/exp.py [hunk 15]
```diff
@@ -280,3 +288,162 @@ def simplify(self):
...
+            "and parameter frequencies can not be computed as no generator is defined."
+        )
+
+
+class Evolution(Exp):
+
+    r"""Create an exponential operator that defines a generator, of the form :math:`e^{ix\hat{G}}`
+
+    Args:
+        base (~.operation.Operator): The Operator to be used as a Generator, G.
```
Comment (obliviateandsurrender) [inline pennylane/ops/op_math/exp.py:377]:
```suggestion
        base (~.operation.Operator): The operator to be used as a generator, G.
```

File: pennylane/ops/op_math/exp.py [hunk 16]
```diff
@@ -280,3 +288,162 @@ def simplify(self):
...
+        param (float): The evolution parameter, x. This parameter is not expected to have
```
Comment (obliviateandsurrender) [inline pennylane/ops/op_math/exp.py:378]: Do we explicitly check this, cast it, or raise any warnings about it?

File: pennylane/ops/op_math/exp.py [hunk 17]
```diff
@@ -280,3 +288,162 @@ def simplify(self):
...
+            any complex component.
+
+    Returns:
+       :class:`Evolution`: A :class`~.operation.Operator` representing an operator exponential of the form exp(ixG),
```
Comment (obliviateandsurrender) [inline pennylane/ops/op_math/exp.py:382]:
```suggestion
       :class:`Evolution`: A :class`~.operation.Operator` representing an operator exponential of the form :math:`e^{ix\hat{G}}`,
```

File: pennylane/ops/op_math/exp.py [hunk 18]
```diff
@@ -280,3 +288,162 @@ def simplify(self):
...
+       where x is real.
+
+    **Use Details**
```
Comment (obliviateandsurrender) [inline pennylane/ops/op_math/exp.py:385]:
```suggestion
    **Usage Details**
```

File: pennylane/ops/op_math/exp.py [hunk 19]
```diff
@@ -280,3 +288,162 @@ def simplify(self):
...
+
+    In contrast to the general Exp class, the Evolution operator is constrained to a single trainable
```
Comment (obliviateandsurrender) [inline pennylane/ops/op_math/exp.py:387]:
```suggestion
    In contrast to the general :class:`~.Exp` class, the Evolution operator :math:`e^{ix\hat{G}}` is constrained to have a single trainable
```

File: pennylane/ops/op_math/exp.py [hunk 20]
```diff
@@ -280,3 +288,162 @@ def simplify(self):
...
+    parameter, x. Any parameters contained in the base operator are not trainable. This allows the operator
+    to be differentiated with regard to the evolution parameter. Defining a mathematically identical operator
+    using the Exp class will be incompatible with a variety of PennyLane functions that require only a single
```
Comment (obliviateandsurrender) [inline pennylane/ops/op_math/exp.py:390]:
```suggestion
    using the :class:`~.Exp` class will be incompatible with a variety of PennyLane functions that require only a single
```

File: tests/ops/op_math/test_exp.py [hunk 21]
```diff
@@ -577,3 +581,123 @@ def test_draw_integration(self):
...
+        with pytest.raises(GeneratorUndefinedError):
+            op.generator()
+
+    def test_generator_is_base_operator(self):
+        """Test that generator is base operator"""
+        base_op = qml.PauliX(0)
+        op = Exp(base_op, 1j)
+        assert op.base == op.generator()
+
+    def test_parameter_frequencies(self):
```
Comment (obliviateandsurrender) [inline tests/ops/op_math/test_exp.py:607]: Might be worth adding the counterexample given by Jay to the tests for testing `qml.simplify` case.

File: tests/ops/op_math/test_exp.py (hunk 12 above)
Comment (obliviateandsurrender) [reply tests/ops/op_math/test_exp.py:586]: +1

**Review by obliviateandsurrender — CHANGES_REQUESTED** [review]
Great work @lillian542, just left some suggestions and comments.

File: pennylane/ops/op_math/exp.py (hunk 9 above)
Comment (lillian542) [reply pennylane/ops/op_math/exp.py]: This is identical to the method on `Operation`, and it has now been changed so it just inherits it from `Operation`.

File: pennylane/ops/op_math/exp.py (hunk 8 above)
Comment (lillian542) [reply pennylane/ops/op_math/exp.py]: This is identical to the method on `Operation`, and it has now been changed so it just inherits it from `Operation`, so this is gone from `Exp` now.

File: pennylane/ops/op_math/exp.py (hunk 13 above)
Comment (lillian542) [reply pennylane/ops/op_math/exp.py]: This method is identical to the method on `Operation`, and it has now been changed so it just inherits it from `Operation`, so this is gone from `Exp` now.

File: pennylane/ops/op_math/exp.py (hunk 14 above)
Comment (lillian542) [reply pennylane/ops/op_math/exp.py]: This method is identical to the method on `Operation`, and it has now been changed so it just inherits it from `Operation`, so this is gone from `Exp` now.

File: tests/ops/op_math/test_exp.py (hunk 21 above)
Comment (lillian542) [reply tests/ops/op_math/test_exp.py:622]: done

File: doc/releases/changelog-dev.md [hunk 22]
```diff
@@ -212,6 +212,30 @@
...
+      Evolution(qml.PauliX(0), -.5 * phi)
+      return qml.expval(qml.PauliZ(0))
+  ```
+  
+  If we run this circuit, we will get the following output
+
+  ```pycon
+  >>> phi = np.array(1.2)
+  >>> qml.grad(circuit)(phi)
+  -0.9320390495504149
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply doc/releases/changelog-dev.md:237]: Include an example of the circuit execution itself

File: pennylane/ops/op_math/exp.py [hunk 23]
```diff
@@ -276,6 +328,116 @@ def pow(self, z):
...
+        for op in [self, qml.simplify(self)]:
+            if op.base.is_hermitian and not np.real(op.coeff):
+                return op.base
+        raise GeneratorUndefinedError(
+            f"Exponential with coefficient {self.coeff} and base operator {self.base} does not have a generator."
+        )
+
+
+class Evolution(Exp):
+
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/ops/op_math/exp.py:361]: remove empty line

File: pennylane/ops/op_math/exp.py [hunk 24]
```diff
@@ -276,6 +328,116 @@ def pow(self, z):
...
+            U(\phi) = e^{i\phi (0.5 Y + Z\otimes X)}
+
+        we get the generator
+
+        >>> U.generator()
+          (0.5) [Y0]
+        + (1.0) [Z0 X1]
+
+        """
+        for op in [self, qml.simplify(self)]:
```
Comment (Jaybsoni) [inline pennylane/ops/op_math/exp.py:352]: I don't think we should be simplifying without telling the user this is happening, It could be very expensive. Could we have a kwarg to determine if we simplify first or not?

File: tests/ops/op_math/test_exp.py [hunk 25]
```diff
@@ -399,6 +405,34 @@ def test_simplify(self):
...
+        """Tests that when simplification of the base results in an SProd,
+        the scalar is included in the coeff rather than the base"""
+        base = qml.s_prod(2, qml.op_sum(qml.PauliX(0), qml.PauliX(0)))
+        op = Exp(base, 3)
+        new_op = op.simplify()
+
+        assert qml.equal(new_op.base, qml.PauliX(0))
+        assert new_op.coeff == 12
+
+    def test_copy(self):
```
Comment (Jaybsoni) [inline tests/ops/op_math/test_exp.py:418]: Also add an assert to make sure there are really two different objects: 
`assert copied_op is not op`

**Review by Jaybsoni — COMMENTED** [review]
Great work! Left some comments for you to address. 

Just one suggestion, since this operation doesn't have a top level constructor (eg. for `Sum` class we have a top level `qml.op_sum`), I think it might be worth while to just make it available top level. This would require adding an import in the PennyLane init file with `from ops import Evolution` and would unlock the `qml.Evolution()` UI 

You should confirm with Tom or Josh to make sure this is actually what we want. Otherwise I think it looks great!

File: pennylane/ops/op_math/exp.py (hunk 24 above)
Comment (obliviateandsurrender) [reply pennylane/ops/op_math/exp.py:352]: Do we really need the kwarg? How about raising a warning to make the user aware? 🤔

```
if self.base.is_hermitian and not np.real(self.coeff):
    return self.base

op = qml.simplify(self)
if op.base.is_hermitian and not np.real(op.coeff):
    warnings.warn("Generator has been computed for the simplified form of Operation")
    return op.base

raise GeneratorUndefinedError(
    f"Exponential with coefficient {self.coeff} and base operator {self.base} does not have a generator."
)
```

File: pennylane/ops/op_math/exp.py [hunk 26]
```diff
@@ -180,6 +203,35 @@ def has_decomposition(self):
...
+        Any subsequent call of this method will toggle between the original
+        operation and the inverse of the operation.
+
+        Returns:
+            :class:`Operator`: operation to be inverted
+        """
+        raise NotImplementedError(
+            f"Setting the inverse of {type(self)} is not implemented. "
+            f"Use qml.adjoint or qml.pow instead."
+        )
```
[the remark this answers is not in the document]
Comment (lillian542) [reply pennylane/ops/op_math/exp.py:233]: They are overwritten because we don't want to inherit them - the base implementation allows you to invert, but these raise an error and tell you to use `adjoint` or `pow` if you try (and return False if you ask op.inverse).

File: pennylane/operation.py [hunk 27]
```diff
@@ -1335,7 +1335,7 @@ def parameter_frequencies(self):
                 eigvals = qml.eigvals(gen)
 
             eigvals = tuple(np.round(eigvals, 8))
-            return qml.gradients.eigvals_to_frequencies(eigvals)
+            return [qml.gradients.eigvals_to_frequencies(eigvals)]
```
[the remark this answers is not in the document]
Comment (lillian542) [reply pennylane/operation.py:1338]: This is something that was added when getting it to work for taking the gradient. It makes the format consistent with how` parameter_frequencies` are defined in f.ex. the `parametric_ops`, where the operators don't call this function but instead have their parameter frequencies defined as a list of tuples. It's also what the docstring suggests the format of the output should be. 

I don't think this was actually related to the issue, I just tried it and removing the brackets doesn't break anything when calculating the gradient via parameter shift. But it makes the output format consistent with both the docstring, and the operators with pre-defined parameter frequencies.

File: pennylane/ops/qubit/hamiltonian.py [hunk 28]
```diff
@@ -393,6 +393,7 @@ def simplify(self):
         self._wires = qml.wires.Wires.all_wires([op.wires for op in self.ops], sort=True)
         # reset grouping, since the indices refer to the old observables and coefficients
         self._grouping_indices = None
+        return self
```
[the remark this answers is not in the document]
Comment (lillian542) [reply pennylane/ops/qubit/hamiltonian.py:396]: In general the `simplify` functions end with returning the simplified operator; as far as I know this was the only exception. Simplifying `SymbolicOps`, including `Exp` and `Evolution`, requires getting the simplified version of the `base`. Without this line, lines 331 and 440 of `exp.py` receive `None` as the simplified base, and then subsequently return an `AttributeError` when trying to use it.

I'm not sure that it is consistent across the various operators whether `simplify` also simplifies the current operator, or creates and returns a new, simplified operator while leaving the current operator untouched. Possibly something we should look into and standardize if inconsistent.

[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/ops/qubit/hamiltonian.py:396]: I think we want to make sure that it is always returning a new operator. I remember `Hamiltonian` being different for some reason. As long as the tests pass I guess this change should be harmless.

File: pennylane/ops/op_math/exp.py [hunk 29]
```diff
@@ -144,14 +146,35 @@ def __repr__(self):
...
+
     @property
     def data(self):
-        return [[self.coeff], self.base.data]
+        return self._data
 
-    @data.setter
-    def data(self, new_data):
-        self.coeff = new_data[0][0]
-        self.base.data = new_data[1]
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/ops/op_math/exp.py:154]: The setter was causing problems.  

When the data is changed, PL tends to use `op.data[i] = new_value`, which bypasses the setter. It just mutates the returned list. I'm amazed this wasn't a problem earlier. But basically, we need to reconsider how we deal with operator parameters and data, but this is a good enough quick fix.

File: pennylane/ops/op_math/exp.py (hunk 24 above)
Comment (Jaybsoni) [reply pennylane/ops/op_math/exp.py:352]: I don't think we should use a warning, because there is no way the user can avoid it (i.e there is no corrective behaviour for the warning except not calling the method). I think the kwarg makes sense, otherwise we can also not simplify, and catch any errors that are caused due to not simplifying, then raise an error telling users to try simplifying first.

File: pennylane/ops/op_math/exp.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (lillian542) [reply pennylane/ops/op_math/exp.py]: I've added a data setter on `Exp` and on `Evolution`. I think the `Exp` one is pretty non-controversial, but I'd like some feedback on the one in `Evolution`. Right now it lets you set either just the primary parameter, or set both the primary parameter and the data for the base. This means it can take an argument that is of a different shape than the operator data:

```
h = qml.PauliX(0) + qml.PauliY(1)
op = Evolution(h, 2)

op.data
>>> [2]
op.base.data
>>> [1, 1]
```

```
op.data = [3]
op.data
>>> [3]
```

```
op.data = [[4], [2, 2]]
op.data
>>> [4]
op.base.data
>>> [2, 2]
```

Is it weird? Maybe you shouldn't be able to set the base data directly through Evolution.data?

File: pennylane/operation.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/operation.py:1338]: Ahh ok awesome! nice catch!

File: pennylane/ops/op_math/exp.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (lillian542) [reply pennylane/ops/op_math/exp.py]: Ah, okay. I just saw Christina's comment. Should I remove them again?

I did notice this morning that setting data on Hamiltonian doesn't actually change the coefficients, only the data list. So if you simplify it after changing the data, for example, it uses the initial data, not the new data.

File: tests/ops/op_math/test_exp.py [hunk 30]
```diff
@@ -399,6 +405,34 @@ def test_simplify(self):
...
+        """Tests making a copy."""
+        op = Exp(qml.CNOT([0, 1]), 2)
+        copied_op = op.__copy__()
+
+        assert qml.equal(op.base, copied_op.base)
+        assert op.wires == copied_op.wires
+        assert op.data == copied_op.data
+        assert op.hyperparameters.keys() == copied_op.hyperparameters.keys()
```
[the remark this answers is not in the document]
Comment (lillian542) [reply tests/ops/op_math/test_exp.py:426]: Here `qml.equal` is comparing only the bases of the two operators, not the operators themselves, so it's compared `op.base.data` and `op.base.hyperparameters`, but those aren't the same as `op.data` and `op.hyperparameters`.
 
 Wires for the base and the operator should be identical for `Exp` and `Evolution` though, I'll get rid of that line.

File: pennylane/ops/op_math/exp.py (hunk 24 above)
Comment (lillian542) [reply pennylane/ops/op_math/exp.py]: I guess the motivating example for this was if the coefficient is real and the base operator has an imaginary component. @Jaybsoni 's example was `base = qml.s_prod(1j, qml.Identity(wires=0))` and `coeff=1`. In that case, if you simplify, you get a correct generator, but if you just check `.is_hermitian` and `not np.real`, in the original formulation of the function you got a `GeneratorUndefinedError`. In this case, we wouldn't want it to proceed with `op.base` as a generator without first simplifying and moving `1j` into the coefficient. 

For the Exp class I've put it back to just raising an error if it doesn't seem to obviously fit the pattern the function is expecting, and suggested in the error message that simplifying the operator might help. I've added a generator function on Evolution that only raises and warning and then proceeds, since the idea with that class is we assume they intentionally defined the base operator to be a generator, unlike if you define an Exp.

Does that seem reasonable?

**Review by obliviateandsurrender — COMMENTED** [review]
Leaving some suggestions. I might not be able to take a look again due to limited internet availability but please feel free to get this merged once you have gone through them and carefully checked the rendering for the docs!

**Review by obliviateandsurrender — APPROVED** [review]

File: pennylane/ops/op_math/exp.py [hunk 31]
```diff
@@ -276,6 +328,115 @@ def pow(self, z):
...
+        """
+        for op in [self, qml.simplify(self)]:
+            if op.base.is_hermitian and not np.real(op.coeff):
+                return op.base
+        raise GeneratorUndefinedError(
+            f"Exponential with coefficient {self.coeff} and base operator {self.base} does not have a generator."
+        )
+
+
+class Evolution(Exp):
```
[the remark this answers is not in the document]
Comment (lillian542) [reply pennylane/ops/op_math/exp.py:362]: Messaged Tom and Josh about this, for now it should just be in ops_math, and a top-level constructor can be considered later.

File: pennylane/ops/op_math/exp.py (hunk 24 above)
Comment (Jaybsoni) [reply pennylane/ops/op_math/exp.py]: Perfect thanks !

**Review by Jaybsoni — APPROVED** [review]
Awesome work!
